### PR TITLE
feat!: add Project.commitGenerated prop

### PR DESCRIPTION
**Pull request by skyrpex**
By default, files managed by projen are commited to the repository. That's not entirely necessary in some cases, and this commit allows customizing this behavior. By providing the Project.defaultCommitManagedFiles prop, we can tell projen whether to commit the files or not (if the File.commited prop is omitted).

The motivation for this change is to have cleaner repositories while maintaining the same developer experience.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/project.ts [hunk 1]
```diff
@@ -94,6 +94,13 @@ export interface ProjectOptions {
    * @default - default options
    */
   readonly renovatebotOptions?: RenovatebotOptions;
+
+  /**
+   * Whether to commit the managed files by default.
+   *
+   * @default true
+   */
+  readonly defaultCommitManagedFiles?: boolean;
```
Comment (eladb) [inline src/project.ts:103]:
```suggestion
  readonly commitGenerated?: boolean;
```

Maybe simpler name?

**Review by eladb — CHANGES_REQUESTED** [review]
There are some files, such as GitHub workflows for example, that MUST be committed. I suspect we need to explicitly indicate that for these files when this option is set to false. Otherwise it's not really usable.

The way I would handle this is add a _required_ prop to FileBase and let the compiler guide you through all the possible files and make an intentional decision whether it must be committed or not.

File: src/project.ts (hunk 1 above)
Comment (skyrpex) [reply src/project.ts]: `commitGenerated` it is.

**Comment by skyrpex** [comment]
> There are some files, such as GitHub workflows for example, that MUST be committed. I suspect we need to explicitly indicate that for these files when this option is set to false. Otherwise it's not really usable.
> 
> The way I would handle this is add a _required_ prop to FileBase and let the compiler guide you through all the possible files and make an intentional decision whether it must be committed or not.

Right, these files need to be committed. The prop you're talking about, do you want it on projen, or are you suggesting it as a temporary means to fix the files that should be truly committed?

**Review by mrgrain — DISMISSED** [review]
To reflect the current status of this PR

File: src/ignore-file.ts [hunk 2]
```diff
@@ -5,7 +5,10 @@ export class IgnoreFile extends FileBase {
   private readonly _patterns = new Array<string>();
 
   constructor(project: Project, filePath: string) {
-    super(project, filePath, { editGitignore: filePath !== ".gitignore" });
+    super(project, filePath, {
+      editGitignore: filePath !== ".gitignore",
+      committed: filePath !== ".gitignore" ? false : undefined,
```
Comment (eladb) [inline src/ignore-file.ts:10]: I believe .gitignore must be committed, no?

File: test/__snapshots__/integ.test.ts.snap [hunk 3]
```diff
@@ -5072,7 +5068,7 @@ jobs:
 !/.github/workflows/pull-request-lint.yml
 !/package.json
 !/LICENSE
-!/.npmignore
```
Comment (Chriscbr) [inline test/__snapshots__/integ.test.ts.snap:5075]: Is this a regression?

**Review by Chriscbr — COMMENTED** [review]
Unit test?

File: src/dependencies.ts [hunk 4]
```diff
@@ -59,6 +59,7 @@ export class Dependencies extends Component {
       new JsonFile(project, Dependencies.MANIFEST_FILE, {
         omitEmpty: true,
         obj: () => this.toJson(),
+        committed: true,
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply src/dependencies.ts:62]: I thought this one file was necessary for projen to maintain some sort of state, but seeing the contents inside it, it doesn't seem required.

File: src/ignore-file.ts (hunk 2 above)
Comment (skyrpex) [reply src/ignore-file.ts:10]: `.gitignore` is committed while the rest of the ignore files aren't, although the logic is hard follow. Maybe a small refactor and some docs will help.

File: src/dependencies.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (eladb) [reply src/dependencies.ts:62]: It' actually not used for anything (besides @MarkMcCulloh's IDE plugin)

File: test/__snapshots__/integ.test.ts.snap (hunk 3 above)
Comment (skyrpex) [reply test/__snapshots__/integ.test.ts.snap:5075]: AFAIK `.npmignore` isn't needed at the repository level, isn't it? For any other operation, it'll be created by projen (eg, the release task).

File: src/github/mergify.ts [hunk 5]
```diff
@@ -77,6 +77,7 @@ export class Mergify extends Component {
           queue_rules: this.queues,
           pull_request_rules: this.rules,
         },
+        committed: true,
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply src/github/mergify.ts:80]: Something like `// Mergify needs to read the file from the repository in order to work.` ?

[the remark this answers is not in the document]
Comment (eladb) [reply src/github/mergify.ts:80]: Yes

File: test/__snapshots__/integ.test.ts.snap (hunk 3 above)
Comment (skyrpex) [reply test/__snapshots__/integ.test.ts.snap:5075]: Yes, didn't get the logic for `IgnoreFile.commited` well.

File: src/ignore-file.ts (hunk 2 above)
Comment (skyrpex) [reply src/ignore-file.ts:10]: Yes, got it all wrong.

File: src/github/pr-template.ts [hunk 6]
```diff
@@ -24,6 +24,7 @@ export class PullRequestTemplate extends TextFile {
           ? options.lines
           : ["Fixes #"],
       marker: false,
+      committed: true,
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply src/github/pr-template.ts:28]: Done.

File: src/github/workflows.ts [hunk 7]
```diff
@@ -83,6 +83,7 @@ export class GithubWorkflow extends Component {
         `.github/workflows/${name.toLocaleLowerCase()}.yml`,
         {
           obj: () => this.renderWorkflow(),
+          committed: true,
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply src/github/workflows.ts:87]: Done.

File: src/gitlab/configuration.ts [hunk 8]
```diff
@@ -156,6 +156,7 @@ export class CiConfiguration extends Component {
         : `.gitlab/ci-templates/${name.toLocaleLowerCase()}.yml`;
     this.file = new YamlFile(this.project, this.path, {
       obj: () => this.renderCI(),
+      committed: true,
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply src/gitlab/configuration.ts:160]: Done.

File: src/gitpod.ts [hunk 9]
```diff
@@ -320,6 +320,7 @@ export class Gitpod extends Component implements IDevEnvironment {
     new YamlFile(this.project, GITPOD_FILE, {
       obj: this.config,
       omitEmpty: true,
+      committed: true,
```
[the remark this answers is not in the document]
Comment (skyrpex) [reply src/gitpod.ts:324]: Done.

**Review by skyrpex — COMMENTED** [review]
Added some explanations to why some files must be committed, and also added some unit tests.

I haven't yet managed to ignore `.projen/deps.json` file if `commitGenerated` is false. Seems like it isn't reading the value of `project.commitGenerated` properly.

**Review by eladb — APPROVED** [review]

**Comment by mrgrain** [comment]
@skyrpex Once you've updated the branch and resolved the conflicts I think we are good to go. :shipit:

File: src/dependencies.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (MarkMcCulloh) [reply src/dependencies.ts]: It is used for state, it's the default method to decide when to delete files that should no longer exist. It's not required per-say (cleanup will fallback to previous logic), but it is used

**Comment by skyrpex** [comment]
@mrgrain rebased and pushed! 🤞🏻

**Review by mrgrain — APPROVED** [review]
Approved to ship

**Pull request merged**